### PR TITLE
randomize pitch black defense times

### PR DESCRIPTION
**Pull request by ArneBab**
wait up to two hours before starting the defense, scheduling the next check for a random time on the next day.

### CODE REVIEW

**Comment by Thynix** [comment]
Are you hoping only for a review of that the implementation does well what it intends to, or are you also looking for review of if doing so is a good idea?

**Review by Thynix — APPROVED** [review]
This certainly seems to do what it intends to in a straightforward way, sure. Whether it makes sense to do that, I don't think I can evaluate.

File: src/freenet/node/LocationManager.java [hunk 1]
```diff
@@ -193,11 +195,12 @@ public void run() {
 
             @Override
             public void run() {
-                node.ticker.queueTimedJob(this, PITCH_BLACK_MITIGATION_FREQUENCY_ONE_DAY);
+                LocalDateTime now = LocalDateTime.now(clockForTesting);
```
Comment (Thynix) [inline src/freenet/node/LocationManager.java:198]: Is this clock not just for running (unit?) tests? Why use it at runtime?

File: src/freenet/node/LocationManager.java [hunk 2]
```diff
@@ -246,7 +249,21 @@ public void run() {
...
+        }, (int) (node.fastWeakRandom.nextFloat() * PITCH_BLACK_MITIGATION_STARTUP_DELAY));
+    }
+
+    /** @return the millis to wait until the next pitch black check: tomorrow and at least 12 hours in the future. */
+    private long getNextPitchBlackMitigationDelayMillisecondsTomorrow(LocalDateTime now) {
+        return Math.max(HOURS.toMillis(12), getMillisUntilRandomTimeTomorrow(now));
+    }
+
+    private long getMillisUntilRandomTimeTomorrow(LocalDateTime now) {
+        LocalDateTime tomorrowHourUTC = now
```
Comment (Thynix) [inline src/freenet/node/LocationManager.java:261]: Why call it hour? That makes it sound like other components will be discarded. What about something like `tomorrowTime`?

**Comment by ArneBab** [comment]
> Are you hoping only for a review of that the implementation does well what it intends to, or are you also looking for review of if doing so is a good idea?

I’m hoping for a review of that the implementation does well what it intends. This is a result of a discussion with nextgens, so I’m pretty sure that it is a good idea to do this.

File: src/freenet/node/LocationManager.java (hunk 1 above)
Comment (ArneBab) [reply src/freenet/node/LocationManager.java:198]: It’s actually a system-clock running at UTC that can be replaced for testing, so the naming was not that great. I renamed it to systemClockUTC now.

Thank you!

File: src/freenet/node/LocationManager.java (hunk 2 above)
Comment (ArneBab) [reply src/freenet/node/LocationManager.java:261]: `tomorrowTime` sounds much better — thank you!

**Comment by ArneBab** [comment]
@Thynix I fixed the naming — can you have one more look at the change? (the last commit: https://github.com/freenet/fred/pull/747/commits/91e34c91d539e31969ac2368587f4e2f80c37207 )

**Pull request merged**